### PR TITLE
Implement the original PQL without support for stochastic environments

### PR DESCRIPTION
**Pull request by wilrop**
The implementation of PQL that currently exists in this repository is not as close to the original algorithm as we would like. Specifically, the algorithm that is on the main branch now has been adapted to work for stochastic environments. While the original paper does mention how PQL should be adapted for stochastic environments, and these changes are implemented on the main branch, to make it work in practice I also had to add a number of non-standard tricks. As this repo is meant to host established "baselines", I believe it is better if we remove the PQL algorithm that is currently in the main branch and use the one that I contribute in this pull request.

In the version that I contribute here, I have exactly replicated the algorithm as mentioned in the original paper. Note that this means it will only work for deterministic environments. In addition, I have also added the two action scoring methods that were explained in the paper, namely hypervolume and Pareto cardinality.

I have, for now, left the PQL algorithm that used to be on the main branch in a different file, named ``pql_stochastic.py``, but I would honestly be in favour of removing it.

### CODE REVIEW

File: morl_baselines/multi_policy/pareto_q_learning/pql.py [hunk 1]
```diff
@@ -1,312 +1,244 @@
...
-            values = []
-            for action in range(self.num_actions):
-                q_set = self.calc_q_set(state, action)
-                q_set = get_non_dominated(q_set)
-                value = self.perf_indic(q_set)
-                values.append(value)
-            action = np.random.choice(np.argwhere(values == np.max(values)).flatten())
-            return action
+        q_sets = [self.get_q_set(state, action) for action in range(self.num_actions)]
+        action_scores = [self.hypervolume(-1 * np.array(list(q_set))) for q_set in q_sets]
```
Comment (ffelten) [inline morl_baselines/multi_policy/pareto_q_learning/pql.py:105]: I guess you can use the hypervolume function in `common` which does the inversion for you.

Comment (wilrop) [reply morl_baselines/multi_policy/pareto_q_learning/pql.py:105]: I thought about doing that, but due to the implementation in ```common```, that way every time you compute the hypervolume a new ```HV``` object is created. I think that will be much slower.

File: morl_baselines/multi_policy/pareto_q_learning/pql.py [hunk 2]
```diff
@@ -1,312 +1,244 @@
...
-                next_state = self.flatten_observation(next_state)
-                self.update(state, action, next_state, r)
+                action = self.select_action(state, score_func)
+                next_state, reward, terminated, truncated, _ = self.env.step(action)
+                next_state = int(np.ravel_multi_index(next_state, self.env_shape))
+
+                self.counts[state, action] += 1
+                self.non_dominated[state][action] = self.calc_non_dominated(next_state)
+                self.avg_reward[state, action] += (reward - self.avg_reward[state, action]) / self.counts[state, action]
+                self.epsilon = max(self.final_epsilon, self.epsilon * self.epsilon_decay)
```
Comment (ffelten) [inline morl_baselines/multi_policy/pareto_q_learning/pql.py:188]: Epsilon is decayed exponentially in the original paper.
image

File: morl_baselines/multi_policy/pareto_q_learning/pql.py (hunk 1 above)
Comment (ffelten) [reply morl_baselines/multi_policy/pareto_q_learning/pql.py:105]: Might be worth trying it. Or modify the common implementation to return the HV object ? :-)

File: morl_baselines/multi_policy/pareto_q_learning/pql.py (hunk 2 above)
Comment (wilrop) [reply morl_baselines/multi_policy/pareto_q_learning/pql.py:188]: I do that every episode, so it is exponential. For example, in episode 5, epsilon is already: ```epsilon = decay * decay * decay * decay * epsilon```.

File: morl_baselines/multi_policy/pareto_q_learning/pql.py (hunk 1 above)
Comment (wilrop) [reply morl_baselines/multi_policy/pareto_q_learning/pql.py]: I now use the implementation from common

File: morl_baselines/multi_policy/pareto_q_learning/pql.py (hunk 2 above)
Comment (wilrop) [reply morl_baselines/multi_policy/pareto_q_learning/pql.py]: I did notice that I updated epsilons inside single episodes, so I took that out and now do it only after having concluded a full episode.

File: morl_baselines/multi_policy/pareto_q_learning/pql_stochastic.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+            "novec": self.novec,
+        }
+
+    def flatten_observation(self, obs):
+        return int(np.ravel_multi_index(obs, (11, 11)))
+
+    def train(
+        self,
+        iterations: int = 3000,
+        max_timesteps: int = 1000,
```
Comment (LucasAlegre) [inline morl_baselines/multi_policy/pareto_q_learning/pql_stochastic.py:262]: Max. timesteps per episode is controlled in the environment itself. If timestep >= max_timesteps, then the environment returns truncated=True

File: morl_baselines/multi_policy/pareto_q_learning/pql_stochastic.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+            "decimals": self.decimals,
+            "novec": self.novec,
+        }
+
+    def flatten_observation(self, obs):
+        return int(np.ravel_multi_index(obs, (11, 11)))
+
+    def train(
+        self,
+        iterations: int = 3000,
```
Comment (LucasAlegre) [inline morl_baselines/multi_policy/pareto_q_learning/pql_stochastic.py:261]: Maybe rename "iterations" to "episodes" ?

**Review by LucasAlegre — CHANGES_REQUESTED** [review]

**Comment by wilrop** [comment]
I have fixed all remarks and deleted the old implementation of PQL. I think the branch is ready to be pulled into main.

**Review by LucasAlegre — APPROVED** [review]

**Comment by LucasAlegre** [comment]
> I have fixed all remarks and deleted the old implementation of PQL. I think the branch is ready to be pulled into main.

Thanks a lot!

**Pull request merged**